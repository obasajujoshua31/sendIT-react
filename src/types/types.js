export const signUpUser = "SIGN_UP_USER";
export const signInUser = "SIGN_IN_USER";
export const createOrder = "CREATE_ORDER";
export const createOrderSuccess = "CREATE_ORDER_SUCCESS";
export const createOrderFailure = "CREATE_ORDER_FAILURE";
export const viewOrders = "VIEW_ORDERS";
export const viewOrder = "VIEW_ORDER";
export const updateParcel = "UPDATE_PARCELS";
export const updateLocation = "UPDATE_PARCEL_LOCATION";
export const cancelOrder = "CANCEL_ORDER";
export const updateParcelStatus = "UPDATE_PARCEL_STATUS";
export const signInSuccess = "SIGN_IN_SUCCESS";
export const signInFailure = "SIGN_IN_FAILURE";
export const signUpSuccess = "SIGN_UP_SUCCESS";
export const signUpFailure = "SIGN_UP_FAILURE";
export const isLoggedOut = "LOGGED_OUT";
export const isAdmin = "IS_ADMIN";
