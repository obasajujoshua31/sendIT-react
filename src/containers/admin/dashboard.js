import React, { Component } from "react";
import { withRouter } from "react-router-dom";
import Spinner from "react-md-spinner";
import loadUserParcels from "../../actions/parcels/admin/getParcels";
import { connect } from "react-redux";
import organiseOrders from "../../helpers/organise";

import { signInFailure, signUpFailure, isLoggedOut } from "../../types/types";

export class AdminDashboard extends Component {
  constructor(props) {
    super(props);
    this.state = {
      authStatus: signInFailure,
      role: false,
      isLoading: true
    };
  }
  componentDidMount() {
    this.props.loadParcels();
  }

  componentDidUpdate(prevProps) {
    if (this.props !== prevProps) {
      if (this.props.parcelStatus) {
        this.props.history.push("/dashboard");
      }
      if (this.props.userParcels.data) {
        this.setState({ isLoading: false });
        this.setState({ parcels: this.props.userParcels.data });
      }
      if (
        this.props.authStatus === signInFailure ||
        this.props.authStatus === signUpFailure ||
        this.props.authStatus === isLoggedOut
      ) {
        this.props.history.push("/login");
      }
    }
  }

  render() {
    if (this.state.isLoading) {
      return (
        <div className="container">
          <div className="table-container">
            <div className="container-not-found">
              <Spinner />
            </div>
          </div>
        </div>
      );
    }

    let placed = 0;
    let delivered = 0;
    let cancelled = 0;
    let transiting = 0;

    if (Object.keys(this.props.userParcels).length > 0) {
      placed = organiseOrders(this.props.userParcels).placed;
      delivered = organiseOrders(this.props.userParcels).delivered;
      cancelled = organiseOrders(this.props.userParcels).cancelled;
      transiting = organiseOrders(this.props.userParcels).transiting;
    }
    return (
      <React.Fragment>
        <div className="account-wrapper">
          <div className="box-group">
            <div className="container">
              <div className="boxes">
                <div className="box">
                  <span className="content">Delivered Orders</span>
                  <span className="content-summary">{delivered}</span>
                </div>
                <div className="box">
                  <span className="content">Placed Orders</span>
                  <span className="content-summary">{placed}</span>
                </div>
              </div>
              <div className="boxes">
                <div className="box">
                  <span className="content">Cancelled Orders</span>

                  <span className="content-summary">{cancelled}</span>
                </div>
                <div className="box">
                  <span className="content">Transiting Orders</span>
                  <span className="content-summary">{transiting}</span>
                </div>
              </div>
            </div>
          </div>
        </div>
      </React.Fragment>
    );
  }
}
export const mapDispatchToProps = dispatch => ({
  loadParcels: () => dispatch(loadUserParcels())
});

export const mapStateToProps = ({ parcels, users }) => ({
  userParcels: parcels.parcels,
  authStatus: users.singInStatus
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(withRouter(AdminDashboard));
