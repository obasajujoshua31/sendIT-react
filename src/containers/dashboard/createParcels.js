import React, { Component } from "react";
import { connect } from "react-redux";
import { withRouter } from "react-router-dom";
import Spinner from "react-md-spinner";
import createParcel from "../../actions/parcels/createOrders";
import { signInFailure, signUpFailure } from "../../types/types";
import validateParcelCreateForm, {
  isValid
} from "../../validations/parcelValidator";
import loadUserParcels from "../../actions/parcels/getParcels";

export class CreateOrder extends Component {
  constructor(props) {
    super(props);

    this.state = {
      parcelName: "",
      destination: "",
      pickUpLocation: "",
      weight: "",
      weightMetric: "Kg",
      isDeactivated: false,
      formErrors: {
        parcelName: "",
        destination: "",
        pickUpLocation: "",
        weight: ""
      },
      isLoading: true,
      parcels: []
    };
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  componentDidMount = () => {
    this.props.loadParcels();
  };

  componentDidUpdate(prevProps) {
    if (prevProps !== this.props) {
      if (
        this.props.authStatus === signInFailure ||
        this.props.authStatus === signUpFailure
      ) {
        this.props.history.push("/login");
      }
      if (this.props.parcels) {
        this.setState({ isLoading: false });
      }
      if (this.props.createdStatus) {
        this.props.history.push("view-orders");
      }
    }
  }

  handleChange(e) {
    const { name, value } = e.target;
    this.setState({ [name]: value });
    this.setState({ formErrors: validateParcelCreateForm(e, this.state) });
    if (!isValid(this.state)) {
      this.setState({ isDeactivated: true });
    } else {
      this.setState({ isDeactivated: false });
    }
  }

  handleSubmit(e) {
    e.preventDefault();
    this.setState({ isLoading: true });
    if (!isValid(this.state)) {
      return this.setState({ isLoading: false });
    } else {
      this.props.createParcel(this.state);
    }
  }

  render() {
    if (this.state.isLoading) {
      return (
        <div className="container">
          <div className="table-container">
            <div className="container-not-found">
              <Spinner />
            </div>
          </div>
        </div>
      );
    }

    const {
      parcelName,
      weight,
      pickUpLocation,
      destination,
      formErrors,
      isDeactivated
    } = this.state;
    return (
      <React.Fragment>
        <div className="account-wrapper">
          <div className="form-wrapper create-wrapper">
            <form onSubmit={this.handleSubmit} noValidate>
              <h2>Create Parcel</h2>

              <div className="form-oneline">
                <div className="form-group">
                  <label htmlFor="parcelName">Parcel Name</label>
                  <input
                    className={formErrors.parcelName.length > 0 ? "error" : ""}
                    type="text"
                    placeholder="Parcel Name"
                    name="parcelName"
                    value={parcelName}
                    onChange={this.handleChange}
                    required
                  />
                  {formErrors.parcelName.length > 0 && (
                    <span className="errorMsg">{formErrors.parcelName}</span>
                  )}
                </div>
                <div className="form-group">
                  <label htmlFor="parcel Weight">Weight</label>
                  <input
                    className={formErrors.weight.length > 0 ? "error" : ""}
                    type="number"
                    placeholder="Weight"
                    name="weight"
                    value={weight}
                    onChange={this.handleChange}
                    required
                  />
                  {formErrors.weight.length > 0 && (
                    <span className="errorMsg">{formErrors.weight} </span>
                  )}
                </div>
              </div>
              <div className="form-group">
                <label htmlFor="weightMetric">Weight Metric</label>
                <select
                  value={this.state.weightMetric}
                  onChange={this.handleChange}
                  name="weightMetric"
                >
                  <option value="Kg">Kg</option>
                  <option value="g">g</option>
                </select>
              </div>

              <div className="form-group">
                <label htmlFor="pickUpLocation">Pick Up Location</label>
                <input
                  className={
                    formErrors.pickUpLocation.length > 0 ? "error" : ""
                  }
                  type="text"
                  placeholder="Your PickUp Location"
                  name="pickUpLocation"
                  value={pickUpLocation}
                  onChange={this.handleChange}
                  required
                />
                {formErrors.pickUpLocation.length > 0 && (
                  <span className="errorMsg">{formErrors.pickUpLocation}</span>
                )}
              </div>

              <div className="form-group">
                <label htmlFor="destination">Your Destination</label>
                <input
                  className={formErrors.destination.length > 0 ? "error" : ""}
                  type="text"
                  placeholder="Your Destination"
                  name="destination"
                  value={destination}
                  onChange={this.handleChange}
                  required
                />
                {formErrors.destination.length > 0 && (
                  <span className="errorMsg">{formErrors.destination}</span>
                )}
              </div>
              <button hidden={isDeactivated} type="submit">
                Send Parcel
              </button>
            </form>
          </div>
        </div>
      </React.Fragment>
    );
  }
}
export const mapDispatchToProps = dispatch => ({
  loadParcels: () => dispatch(loadUserParcels()),
  createParcel: parcel => dispatch(createParcel(parcel))
});

export const mapStateToProps = ({ users, parcels }) => ({
  authStatus: users.singInStatus,
  createdStatus: parcels.parcelStatus,
  parcels: parcels.parcels
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(withRouter(CreateOrder));
