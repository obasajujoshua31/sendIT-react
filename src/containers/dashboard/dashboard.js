import React, { Component } from "react";
import { connect } from "react-redux";

import { withRouter } from "react-router-dom";
import Spinner from "react-md-spinner";
import loadUserParcels from "../../actions/parcels/getParcels";
import { signInFailure, signUpFailure } from "../../types/types";
import organiseOrders from "../../helpers/organise";

export class Dashboard extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isLoading: true
    };
  }
  componentDidMount() {
    this.props.loadParcels();
  }
  componentDidUpdate(prevProps) {
    if (prevProps !== this.props) {
      if (
        this.props.authStatus === signInFailure ||
        this.props.authStatus === signUpFailure
      ) {
        this.props.history.push("/login");
      }
      this.setState({isLoading: false})
    }
  }

  render() {
    if (this.state.isLoading) {
      return (
        <div className="container">
          <div className="table-container">
            <div className="container-not-found">
              <Spinner />
            </div>
          </div>
        </div>
      );
    }

    let placed = 0;
    let delivered = 0;
    let cancelled = 0;
    let transiting = 0;

    if (Object.keys(this.props.userParcels).length > 0) {
      placed = organiseOrders(this.props.userParcels).placed;
      delivered = organiseOrders(this.props.userParcels).delivered;
      cancelled = organiseOrders(this.props.userParcels).cancelled;
      transiting = organiseOrders(this.props.userParcels).transiting;
    }

    return (
      <React.Fragment>
        <div className="account-wrapper">
          <div className="box-group">
            <div className="container">
              <div className="boxes">
                <div className="box">
                  <span className="content">Delivered Orders</span>
                  <span className="content-summary">{delivered}</span>
                </div>
                <div className="box">
                  <span className="content">Placed Orders</span>
                  <span className="content-summary">{placed}</span>
                </div>
              </div>
              <div className="boxes">
                <div className="box">
                  <span className="content">Cancelled Orders</span>

                  <span className="content-summary">{cancelled}</span>
                </div>
                <div className="box">
                  <span className="content">Transiting Orders</span>
                  <span className="content-summary">{transiting}</span>
                </div>
              </div>
            </div>
          </div>
        </div>
      </React.Fragment>
    );
  }
}

export const mapDispatchToProps = dispatch => ({
  loadParcels: () => dispatch(loadUserParcels())
});

export const mapStateToProps = ({ parcels, users }) => ({
  userParcels: parcels.parcels,
  authStatus: users.singInStatus
});
export default connect(
  mapStateToProps,
  mapDispatchToProps
)(withRouter(Dashboard));
