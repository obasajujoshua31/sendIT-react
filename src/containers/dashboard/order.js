import React, { Component } from "react";
import { withRouter } from "react-router-dom";

class Order extends Component {
  constructor(props) {
    super(props);
  }
  render() {
    const {
      parcel: {
        parcel_id,
        parcel_name,
        destination,
        pick_up_location,
        sent_on,
        status
      },
      onHandleViewOrder
    } = this.props;
    return (
      <React.Fragment>
        <tr onClick={() => onHandleViewOrder(parcel_id)}>
          <th scope="row">{`#${parcel_id}`}</th>
          <td>{parcel_name}</td>
          <td>{pick_up_location}</td>
          <td>{destination}</td>
          <td>{sent_on}</td>
          <td>{status}</td>
        </tr>
      </React.Fragment>
    );
  }
}

export default withRouter(Order);
