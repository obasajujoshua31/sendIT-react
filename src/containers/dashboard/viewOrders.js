import React, { Component } from "react";
import loadUserParcels from "../../actions/parcels/getParcels";
import { connect } from "react-redux";
import { withRouter, Link } from "react-router-dom";
import Spinner from "react-md-spinner";
import {
  signInFailure,
  signUpFailure,
  signInSuccess,
  signUpSuccess
} from "../../types/types";
import Order from "./order";
import EditModal from "../../components/modals/edit";
import ReactModal from "react-modal";

export class ViewOrders extends Component {
  constructor(props) {
    super(props);
    this.state = {
      parcels: [],
      showModal: false,
      parcelId: "",
      isLoading: true
    };
    this.onHandleViewOrder = this.onHandleViewOrder.bind(this);
    this.closeModal = this.closeModal.bind(this);
  }

  onHandleViewOrder(parcelId) {
    this.setState({
      showModal: true,
      parcelId,
      authStatus: signInSuccess || signUpSuccess
    });
  }
  closeModal() {
    this.setState({ showModal: false });
  }
  componentWillMount() {
    if (process.env.NODE_ENV !== "test") {
      ReactModal.setAppElement("#app");
      ReactModal.defaultStyles.overlay.backgroundColor = "cornsilk";
    }
  }

  componentDidMount() {
    this.props.loadParcels(this.props.history);
  }

  componentDidUpdate(prevProps) {
    if (prevProps !== this.props) {
      if (this.props.userParcels) {
        this.setState({ isLoading: false });
        this.setState({ parcels: this.props.userParcels });
      }
      if (
        this.props.authStatus === signInFailure ||
        this.props.authStatus === signUpFailure
      ) {
        this.props.history.push("/login");
      }
    }
  }

  render() {
    if (this.state.isLoading) {
      return (
        <div className="container">
          <div className="table-container">
            <div className="container-not-found">
              <Spinner />
            </div>
          </div>
        </div>
      );
    }
    if (!this.state.parcels.length) {
      return (
        <div className="container">
          <div className="table-container">
            <div className="container-not-found">
              <h1>You are Welcome to SendIT Courier Services</h1>
              <h1>You don't have any parcels yet</h1>
              <h1>
                <Link to="/create-order">Click here to create a Parcel</Link>
              </h1>
            </div>
          </div>
        </div>
      );
    }

    return (
      <React.Fragment>
        <div className="container">
          <div className="table-container">
            <h1
              style={{
                textAlign: "center",
                paddingTop: "20px",
                color: "#806B75"
              }}
            >
              Your Orders
            </h1>
            <table className="table" align="center">
              <thead className="thead-dark">
                <tr>
                  <th scope="col">Parcel Id</th>
                  <th scope="col">Parcel Name</th>
                  <th scope="col">Pick Up Location</th>
                  <th scope="col">Destination</th>
                  <th scope="col">Sent On</th>
                  <th scope="col">Status</th>
                </tr>
              </thead>
              <tbody>
                {this.state.parcels.map(value => {
                  return (
                    <React.Fragment key={value.parcel_id}>
                      <Order
                        key={value.parcel_id}
                        parcel={value}
                        onHandleViewOrder={this.onHandleViewOrder}
                      />
                    </React.Fragment>
                  );
                })}
              </tbody>
            </table>
          </div>
        </div>
        {this.state.showModal && (
          <EditModal
            show={this.state.showModal}
            handleCloseModal={this.closeModal}
            parcelId={this.state.parcelId}
            role={false}
            header={"Change Destination"}
          />
        )}
      </React.Fragment>
    );
  }
}

export const mapDispatchToProps = dispatch => ({
  loadParcels: () => dispatch(loadUserParcels())
});

export const mapStateToProps = ({ parcels, users }) => ({
  userParcels: parcels.parcels,
  authStatus: users.singInStatus,
  role: users.role,
});
export default connect(
  mapStateToProps,
  mapDispatchToProps
)(withRouter(ViewOrders));
