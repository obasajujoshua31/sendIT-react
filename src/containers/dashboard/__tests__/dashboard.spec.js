import React from "react";
import { Dashboard, mapDispatchToProps, mapStateToProps } from "../dashboard";
import { shallow } from "enzyme";
import sinon from "sinon";
import { signInSuccess, signInFailure } from "../../../types/types";

const props = {
  userParcels: [],
  loadParcels: jest.fn(),
  history: {
    push: jest.fn()
  }
};

const mockParcels = {
  data: [
    {
      parcel: {
        status: "PLACED"
      }
    },
    {
      parcel: {
        status: "DELIVERED"
      }
    },
    {
      parcel: {
        status: "TRANSITING"
      }
    },
    {
      parcel: {
        status: "CANCELLED"
      }
    }
  ]
};

const mockStore = {
  users: {
    singInStatus: signInSuccess
  },
  parcels: {
    parcels: ["Beans"]
  }
};
const wrapper = shallow(<Dashboard {...props} />);
describe("Test for User Dashboard", () => {
  it("should render Dashboard component", () => {
    expect(wrapper).toMatchSnapshot();
  });
  it("should dispatch loadparcels", () => {
    const dispatchSpy = sinon.spy();
    const dispatchResult = mapDispatchToProps(dispatchSpy);
    dispatchResult.loadParcels();
    expect(dispatchSpy.callCount).toEqual(1);
  });

  it("should return new object for mapstate to props", () => {
    const stateToPropsResult = mapStateToProps(mockStore);
    expect(stateToPropsResult.userParcels).toEqual(["Beans"]);
    expect(stateToPropsResult.authStatus).toEqual(signInSuccess);
  });

  it("should ", () => {
    const dashboardWrapper = shallow(<Dashboard {...props} />);
    dashboardWrapper.setProps({
      userParcels: mockParcels
    });
  });

  it("should call history.push", () => {
    const dashboardWrapper = shallow(<Dashboard {...props} />);
    dashboardWrapper.setProps({
      authStatus: signInFailure
    });
    expect(dashboardWrapper.instance().props.history.push).toBeCalled();
  });
});
