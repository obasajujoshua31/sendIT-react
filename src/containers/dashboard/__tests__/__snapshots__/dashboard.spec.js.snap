// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test for User Dashboard should render Dashboard component 1`] = `
<div
  className="container"
>
  <div
    className="table-container"
  >
    <div
      className="container-not-found"
    >
      <MDSpinner
        color1="rgb(66, 165, 245)"
        color2="rgb(239, 83, 80)"
        color3="rgb(253, 216, 53)"
        color4="rgb(76, 175, 80)"
        duration={1333}
        size={28}
      />
    </div>
  </div>
</div>
`;
